### ISSUE TITLE
Return null in data for error messages

### ISSUE THREAD

**Issue by mgwidmann** [absinthe-graphql/absinthe#305]
Discussion here:
https://github.com/apollographql/apollo-ios/pull/4#issuecomment-290753694

Here's the JSON response for a custom error message:

```javascript
{
  "errors": [
    {
      "user_message": "No user found for given email.",
      "message": "In field \"signInUser\": Email not found error",
      "locations": [
        {
          "line": 2,
          "column": 0
        }
      ]
    }
  ],
  "data": {}
}
```

The Apollo iOS framework (in the above discussion) is stating that the spec dictates the `data` key should be `null` instead of `{}` when an error is in the response. They claim that for this reason it messes with the logic which prevents the error from actually being returned in the callback provided to the iOS apollo library.

Using absinthe 1.2.5

**Comment by benwilson512** [comment]
Sorry but you're gonna need to provide a summary of what the issue is here.

**Comment by mgwidmann** [comment]
Updated the description. Let me know if there are any other questions.

**Comment by bruce** [comment]
This is correct based on https://facebook.github.io/graphql/#sec-Data.

> If an error was encountered before execution begins, the `data` entry should not be present in the result.
> If an error was encountered during the execution that prevented a valid response, the data entry in the response should be `null`.

I'm pretty sure this is a regression, but I'm not sure when it occurred.

**Comment by benwilson512** [comment]
I'm not sure that this is correct. The error here is a resolution time error on that specific field.

Per https://facebook.github.io/graphql/#sec-Value-Completion the result of the `query {}` operation is the root query type, and then you recursively evaluate fields under that type. Unless an empty object is ipso facto an invalid response, I'm not sure that this constitutes "an error was encountered during the execution that prevented a valid response"

**Comment by benwilson512** [comment]
By way of elaboration, suppose you had two mutation fields in this operation IE:

```
mutation M {
  signInUser(args)
  foo
}
```

The error in `signInUser` is not so problematic that it would have prevented a valid response for `foo`, which would give us a response of:
```
{"data":{"foo":"bar"}, "errors": [...]}
```

If we remove `foo` from the document and from the result we get an empty map data, not null data.

**Comment by bruce** [comment]
I think this is a matter of how you read "prevents a valid response." It's interesting that https://facebook.github.io/graphql/#sec-Data doesn't explicitly talk about empty objects.

This set of tests may be instructive: https://github.com/graphql/graphql-js/blob/master/src/execution/__tests__/nonnull-test.js

Notably, I can't find a single test case in graphql-js that asserts a value is `data: {}`.

**Comment by benwilson512** [comment]
True. At the same time every example of `data: null` involves a top level `non-nullable field throws`.

I think part of the issue here is figuring out what is meant by "preventing a valid response".

**Comment by mgwidmann** [comment]
For what its worth, the Apollo iOS library believes this to be an issue (and ends up having a pretty bad user experience on the client side in this condition)... So if the argument is that its right according to spec we'll need a good argument to convince them to change (someone has to change here...)

**Comment by benwilson512** [comment]
Right yeah. I'm also curious about what other server languages do.

**Comment by bruce** [comment]
FWIW in the short term as a workaround, this is fixable by adding a simple phase to the end of your pipeline.

**Comment by mgwidmann** [comment]
For those finding this issue as well, heres how to add the phase described above:

```elixir
# Add this module anywhere
defmodule MyApp.NullDataForErrorsPhase do
  use Absinthe.Phase

  # Copied from https://github.com/absinthe-graphql/absinthe_plug/blob/master/lib/absinthe/plug.ex#L267
  def pipeline(config, pipeline_opts) do
    config.schema_mod
    |> Absinthe.Pipeline.for_document(pipeline_opts)
    |> Absinthe.Pipeline.insert_after(Absinthe.Phase.Document.CurrentOperation,
      {Absinthe.Plug.Validation.HTTPMethod, method: config.conn_private.http_method})
    |> Absinthe.Pipeline.insert_after(Absinthe.Phase.Document.Result,
      __MODULE__)
  end

  # No data being returned, with error
  def run(%{data: %{}, errors: [%{}] = errors}, _options) do
    {:ok, %{data: nil, errors: errors}}
  end
  def run(blueprint, _options) do
    {:ok, blueprint}
  end
end

# And then where you run your schema
forward "/graphql", Absinthe.Plug,
      schema: MyApp.Schema,
      pipeline: {MyApp.NullDataForErrorsPhase, :pipeline}
```

**Comment by benwilson512** [comment]
So, this is not what graphql-js does. If an inner field throws an error, that field should be null. This does mean that Absinthe has a bug here, we should be returning `{data:{signInUser: null}, errors:[...]}` but we definitely shouldn't be returning `data: null`.

Fixing this bug may however solve your problem.

http://graphql.org/swapi-graphql/?query=query%20%7B%0A%20%20allFilms(first%3A%20-10)%20%7B%0A%20%20%20%20__typename%0A%20%20%7D%0A%7D

**Comment by benwilson512** [comment]
Actually @mgwidmann this is already fixed in 1.3.0. I would recommend upgrading.
